### PR TITLE
Polish flux metrics

### PR DESCRIPTION
**Pull request by simonbasle**
Reviewing commits a9cc14e2e9907d8f7b46c8f0aa678ffed90a1e92, f478ab30c1ba810d97ed31570829adb953406209, 610a77025b131fa66e8d1f27262ac92d372a2bb7, d1d45c69960df03baa1f5f3848d43abacf1d76ed and 6897e3f94041c4a7bf57014312b244f3d6aa1c37 I came across a few polish points summarized below:

 - [x] Use the `Tags::and` combiner for reduce, even though it is only useful for parallel streams (do the right thing, `(a,b)->b` would totally break things if the stream was parallelized)
 - [x] Tags.of(k, v) is even leaner on allocations than Tags.of(Tag.of(k,v))
 - [x] Explicitly comment record* methods are special and can call registry
 - [x] Align all 4 implementations to make registryCandidate non-null in ctor

### CODE REVIEW

File: reactor-core/src/main/java/reactor/core/publisher/MonoMetrics.java [hunk 1]
```diff
@@ -132,6 +131,7 @@ final public void onNext(T t) {
 				return;
 			}
 			done = true;
+			//TODO looks like we don't count onNext: `Mono.empty()` vs `Mono.just("foo")`
 			FluxMetrics.recordOnComplete(commonTags, registry, subscribeToTerminateSample);
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/MonoMetrics.java]: @smaldini shouldn't we increment an `onNext` counter here, in order to differentiate empty case vs valued case? (this is not a regression and potentially a candidate for an `enhancement` follow-up issue)

File: reactor-core/src/main/java/reactor/core/publisher/FluxMetrics.java [hunk 2]
```diff
@@ -298,12 +297,21 @@ static Tags resolveTags(Publisher<?> source, Tags tags, String sequenceName) {
 
 		if (scannable.isScanAvailable()) {
 			return scannable.tags()
-			                .reduce(tags, TAG_ACCUMULATOR, TAG_COMBINER);
+			                //Note the combiner below is for parallel streams, which won't be used
+			                //For the identity, `commonTags` should be ok (even if reduce uses it multiple times)
+			                //since it deduplicates
+			                .reduce(tags, TAG_ACCUMULATOR, Tags::and);
```
Comment (smaldini) [inline reactor-core/src/main/java/reactor/core/publisher/FluxMetrics.java]: Can we use a static Tags::and ?

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/FluxMetrics.java]: done

File: reactor-core/src/main/java/reactor/core/publisher/MonoMetrics.java (hunk 1 above)
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/MonoMetrics.java]: (removed the comment and moved that in a separate issue #1803)

**Pull request merged**